### PR TITLE
feat(vite-plugin-nitro): add support for XML content in API routes

### PR DESCRIPTION
**Pull request by justinrassier**
Add the ability to set content-type as XJML in server routes and to have SSG correctly pre-render the data

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/analogjs/analog/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

- [ ] vite-plugin-angular
- [x] vite-plugin-nitro
- [ ] astro-angular
- [ ] create-analog
- [ ] router
- [ ] platform
- [ ] content
- [ ] nx-plugin
- [ ] trpc

## What is the current behavior?

When defining an API route with a `content-type` other than `applicaton/json`, nitro interprets all responses as `text/html` and renders a `.html` document for the data. If you would like to serve up something like XML for an RSS feed, this doesn't work.

## What is the new behavior?
If you set the headers to `content-type: text/xml` (or anything non-json) it re-proxies the request to allow nitro to interpret the result as intended.

Note: This is a bit wasteful to re-proxy the request, but I was unable to bet nitro to be happy with any other approach. I am totally willing to try other approaches, but this was the most consistent way I could get the desired behavior

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by brandonroberts** [comment]
Nice! What about making it more straightforward and checking if the `reqUrl` ends in `.xml` or the `event.node.req` headers for the `content-type` include `application/json`?

If it does, skip over the `$fetch` and go to the `proxyRequest`.

**Comment by justinrassier** [comment]
> Nice! What about making it more straightforward and checking if the `reqUrl` ends in `.xml` or the `event.node.req` headers for the `content-type` include `application/json`?
> 
> If it does, skip over the `$fetch` and go to the `proxyRequest`.

I can totally check for the `.xml` ending. There is not `content-type` for  the request in this case. 

I'll make that change to the PR to keep it a bit more targeted and simple .The e2e tests should still apply so I'll run those to make sure it still passes.

**Comment by justinrassier** [comment]
PR updated to keep it a bit more targeted and simple. Only looking for `.xml` as the ending of the request URL.

**Comment by brandonroberts** [comment]
Great. For context, the reason the `GET` requests don't get forwarded to the `proxyRequest` is that the response isn't correct and it breaks the JSON response. Needs further investigation there though outside of this PR.

File: apps/docs-app/docs/features/api/overview.md [hunk 1]
```diff
@@ -13,7 +13,45 @@ import { defineEventHandler } from 'h3';
 export default defineEventHandler(() => ({ message: 'Hello World' }));
 ```
 
-API routes are powered by [Nitro](https://nitro.unjs.io). See the Nitro docs for more examples around building API routes.
+### Defining XML Content
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/api/overview.md:16]:
```suggestion
## Defining XML Content
```

File: apps/docs-app/docs/features/api/overview.md [hunk 2]
```diff
@@ -13,7 +13,45 @@ import { defineEventHandler } from 'h3';
...
+
+If you want to create an RSS feed for your site, you can set the `content-type` to be `text/xml` and Analog will correctly serve up the correct content type for the route.
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/api/overview.md:18]:
```suggestion
To create an RSS feed for your site,  set the `content-type` to be `text/xml` and Analog serves up the correct content type for the route.
```

File: apps/docs-app/docs/features/api/overview.md [hunk 3]
```diff
@@ -13,7 +13,45 @@ import { defineEventHandler } from 'h3';
...
+  const feedString = `<?xml version="1.0" encoding="UTF-8"?>
+<rss version="2.0">
+</rss>
+  `;
+  event.node.res.setHeader('content-type', 'text/xml');
+  event.node.res.end(feedString);
+});
+```
+
+**Note:** For SSG content, you can set Analog to prerender an API route to make it available as prerendered content:
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/api/overview.md:34]:
```suggestion
**Note:** For SSG content, set Analog to prerender an API route to make it available as prerendered content:
```

**Comment by justinrassier** [comment]
> Great. For context, the reason the `GET` requests don't get forwarded to the `proxyRequest` is that the response isn't correct and it breaks the JSON response. Needs further investigation there though outside of this PR.

Thanks for the context! With that in mind, then yes, this is way safer just checking `.xml` for the route so it wouldn't break any other expectations!

File: apps/docs-app/docs/features/api/overview.md [hunk 4]
```diff
@@ -13,7 +13,45 @@ import { defineEventHandler } from 'h3';
...
+
+```ts
+// vite.config.ts
+...
+prerender: {
+  routes: async () => {
+    return [
+      ...
+      './api/rss.xml',
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/api/overview.md:43]:
```suggestion
      '/api/rss.xml',
```

File: apps/docs-app/docs/features/api/overview.md [hunk 5]
```diff
@@ -13,7 +13,45 @@ import { defineEventHandler } from 'h3';
...
+      .
+    ];
+  },
+  sitemap: {
+    host: 'https://analog-blog.netlify.app',
+  },
+},
+```
+
+The XML will be available as a static XML document at `/dist/analog/public/api/rss.xml`
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/api/overview.md:54]:
```suggestion
The XML is available as a static XML document at `/dist/analog/public/api/rss.xml`
```

File: apps/blog-app-e2e-cypress/src/e2e/app.cy.ts [hunk 6]
```diff
@@ -4,4 +4,14 @@ describe('blog-app', () => {
   it('should redirect us to /blog', () => {
     cy.url().should('be.equal', `${Cypress.config('baseUrl')}/blog`);
   });
+  it('should serve up HTML for pre-rendered markdown route', () => {
+    cy.visit('/blog/2022-12-27-my-first-post');
+    cy.get('h1').should('contain', 'My First Post');
+  });
+  it('should serve up XML for pre-rendered XML route from vite.config at /api/rss', () => {
```
Comment (brandonroberts) [inline apps/blog-app-e2e-cypress/src/e2e/app.cy.ts:11]:
```suggestion
  it('should serve up XML for pre-rendered XML route from vite.config at /api/rss.xml', () => {
```

**Review by brandonroberts — CHANGES_REQUESTED** [review]
Docs tweaks

**Comment by justinrassier** [comment]
Whoops, sorry for the spam with updates. Thought I would take the docs changes through the UI but ended up creating a lot of noise. Sorry!

😬

**Review by brandonroberts — APPROVED** [review]

**Comment by brandonroberts** [comment]
@allcontributors add @justinrassier for code

**Pull request merged**